plugins/vli/fu-vli-pd-parade-device.c

```diff
@@ -0,0 +1,668 @@
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x92, 0x00, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x93, 0x05, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_set_wp (self, FALSE, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_write_disable (FuVliPdParadeDevice *self, GError **error)
+{
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0xDA, 0x00, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_write_status (FuVliPdParadeDevice *self, guint8 target_status, GError **error)
+{
+	/* Set_WP_High, SPI_WSTS_01, Target_Status, Len_01, Trigger_Write, Set_WP_Low */
+	if (!fu_vli_pd_parade_device_set_wp (self, TRUE, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, 0x01, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, target_status, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x92, 0x01, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x93, 0x05, error))
+		return FALSE;
+	if (!fu_vli_pd_parade_device_set_wp (self, FALSE, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_vli_pd_parade_device_wait_ready (FuVliPdParadeDevice *self, GError **error)
+{
+	gboolean ret = FALSE;
+	guint limit = 100;
+	guint8 buf = 0x0;
+
+	/* Wait_For_SPI_ROM_Until_Not_Busy */
+	for (guint wait_cnt1 = 0; wait_cnt1 < limit; wait_cnt1++) {
+		buf = 0xFF;
+		if (!fu_vli_pd_parade_device_i2c_read (self, self->page_addr,
+						       0x9E, &buf, sizeof(buf),
+						       error))
+			return FALSE;
+		/* Busy_Status, bit[1,0]:Byte_Program, bit[3,2]:Sector_Erase, bit[5,4]:Chip_Erase */
+		if ((buf & 0x0C) == 0) {
+			ret = TRUE;
+			break;
+		}
+	}
+	if (!ret) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_INTERNAL,
+			     "Wait_9e_TOut");
+		return FALSE;
+	}
+
+	/* Wait_SPI_Until_Ststus_Bit_0_Clear */
+	ret = FALSE;
+	for (guint wait_cnt1 = 0; wait_cnt1 < limit; wait_cnt1++) {
+		gboolean ret2 = FALSE;
+
+		/* SPI_RSTS_05, Len_01, Trigger_Read */
+		if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x90, 0x05, error))
+			return FALSE;
+		if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x92, 0x00, error))
+			return FALSE;
+		if (!fu_vli_pd_parade_device_i2c_write (self, self->page_addr, 0x93, 0x01, error))
+			return FALSE;
+
+		/* Wait_SPI_CMD_Done */
+		for (guint wait_cnt2 = 0; wait_cnt2 < limit; wait_cnt2++) {
+			buf = 0xFF;
+			if (!fu_vli_pd_parade_device_i2c_read (self, self->page_addr,
+							       0x93, &buf, sizeof(buf),
+							       error))
+				return FALSE;
+			if ((buf & 0x01) == 0) {
+				ret2 = TRUE;
+				break;
+			}
+		}
+		if (!ret2) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "Wait_93_TOut");
```

Comment:
likewise, not really a user friendly string

Reply:
Fixed.